src/dist/mod.rs

```diff
@@ -319,7 +320,11 @@ impl FromStr for ParsedToolchainDesc {
         if let Some(d) = d {
             Ok(d)
         } else {
-            Err(RustupError::InvalidToolchainName(desc.to_string()).into())
+            let desc_string = desc.to_string();
+            if desc_string.as_str() == "self" {
+                info!("If you meant to modify rustup itself, try run `rustup self`");
+            }
+            Err(RustupError::InvalidToolchainName(desc_string).into())
```

Comment:
I think it would be better to do this near the top of the method? We can just check for `desc == "self"` and return the error immediately, which would save a bunch of work trying to parse.

Reply:
@djc My suggestion is that since `FromStr` is used in many places other than `update` and `uninstall`, this will cause many false positives, and instead we should modify the `value_parser` for these two commands only.

Reply:
But `self` is still invalid in all those places, right -- worst case this will provide a clearer error message?

Reply:
@djc I'm afraid that will make the error message less clear. `rustup install self` doesn't make sense and if the info leads the user to `rustup self` they won't find anything related to `install` there.

Reply:
Sure, but users are presumably also much less likely to invoke `rustup install self`. I still think it would be better to embed this check inside the `FromStr` impl where it is implicitly used in any future code that wants to consume toolchain names instead of remembering that we have to add an extra check everywhere, we can make the error message slightly more generic to make sure it doesn't lose clarity.

Reply:
@djc I still feel uneasy about adding side effects in conversions like `FromStr`... @ChrisDenton What do you think?

Reply:
I would argue that logging is pretty much "unobservable" as side effects go, so while I agree that in general side effects in something like a `FromStr` implementations could be a bit iffy, I think logging is an exception to this.

Would you have problems with adding logging to other "pure" functions?

Reply:
@djc Actually... it's not about purity, it's about the fact that conversions can be used in various places across the codebase, many of which are not easily detectable via call hierarchy graph `rust-analyzer` gives us, and I'm afraid of this potentially adding a log line somewhere unexpected. Since the very nature of this feature is to add a hint, I guess it'd be better to get the hint only when we are explicitly expecting for it.

For example, `rustup uninstall` uses `ResolvableToolchainName`, and its conversions relies on the conversions defined for `PartialToolchainDesc` first, and that of `CustomToolchainName` as a fallback. Let's say we're still using https://github.com/rust-lang/rustup/commit/a743b3adc19ec949761a236fe484b0e34f0f44fb, and it so happens that there's a custom toolchain named `self`, before the fallback happens, the log line is already printed out.